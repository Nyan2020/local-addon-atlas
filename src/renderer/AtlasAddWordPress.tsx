import React, { useState } from 'react';
import {
	AdvancedToggle,
	FlySelect,
	PrimaryButton,
	TextButton,
	Title,
	InputPasswordToggle,
	BasicInput,
} from '@getflywheel/local-components';
import { MultiSite, NewSiteInfo } from '@getflywheel/local';
import { sendIPCEvent, RouteComponentProps } from '@getflywheel/local/renderer';
import { regexPatterns } from '../constants';

interface IProps extends RouteComponentProps {
	siteSettings: NewSiteInfo
}

export const AtlasAddWordPress = (props: IProps): JSX.Element => {
	const [adminUsername, setUsername] = useState('');
	const [adminPassword, setPassword] = useState('');
	const [adminEmail, setEmail] = useState('dev-email@flywheel.local');
	const [multisite, setMultisite] = useState<MultiSite | 'no'>('no');

	const isValid = adminUsername && adminPassword && adminEmail.match(regexPatterns.email);

	const createSiteFromBlueprint = () => {
		if (isValid) {
			sendIPCEvent('addSite', {
				newSiteInfo: props.siteSettings,
				wpCredentials: {
					adminUsername,
					adminPassword,
					adminEmail,
				},
				goToSite: true,
			});
		}
	};

	return (
		<div className="AddSiteContent">
			<Title size="l" container={{ margin: 'l 0' }}>Set up WordPress</Title>

			<div className="Inner">
				<div className="AtlasAddWordPress_Inputs">
					<div>
						<label htmlFor="adminUsername">WordPress username</label>
						<BasicInput
							invalid={!adminUsername}
							invalidMessage={'WordPress admin username is invalid.'}
							id="adminUsername"
							value={adminUsername}
							onChange={(e) => setUsername(e.target.value)}
						/>
					</div>

					<div>
						<label htmlFor='adminPassword'>WordPress password</label>
						<InputPasswordToggle
							invalid={!adminPassword}
							invalidMessage={'WordPress admin password is invalid.'}
							className="TID_AddSiteWordPress_PasswordToggle_AdminPassword"
							id="adminPassword"
							value={adminPassword}
							onChange={(e) => setPassword(e.target.value)}
						/>
					</div>

					<div>
						<label htmlFor="adminEmail">WordPress email</label>
						<BasicInput
							invalid={!adminEmail.match(regexPatterns.email)}
							invalidMessage={'WordPress admin email is invalid.'}
							id="adminEmail"
							value={adminEmail}
							onChange={(e) => setEmail(e.target.value)}
						/>
					</div>
				</div>

				<AdvancedToggle className="TID_AddSiteWordPress_AdvancedToggle_AdvancedOptions">
					<div className="FormRow FormRow__Third FormRow__Center">
						<div className="FormField">
							<label>Is this a WordPress Multisite?</label>
							<FlySelect
								value={multisite}
								options={{
									'no': 'No',
									'ms-subdir': 'Yes – Subdirectory',
									'ms-subdomain': 'Yes – Subdomain',
								}}
								onChange={setMultisite}
							/>
						</div>
					</div>
				</AdvancedToggle>
			</div>

			<PrimaryButton
				disabled={!isValid}
				className="TID_AddSiteWordPress_Button_Continue Continue"
				onClick={createSiteFromBlueprint}
			>
				Create site from Blueprint
			</PrimaryButton>
			<TextButton
				className="GoBack"
				onClick={() => props.history.goBack()}
			>
				Go back
			</TextButton>
		</div>
	);
};
