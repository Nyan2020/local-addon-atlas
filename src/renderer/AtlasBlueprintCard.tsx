import React from 'react';
import { Card, Divider, Text } from '@getflywheel/local-components';

type AtlasBlueprintCardProps = {
	thumbnailSrc: string;
	title: string;
	byline: string;
	excerpt: string;
	previewHref: string;
	repoHref: string;
	detailsHref: string;
};

const AtlasBlueprintCard: React.FC<AtlasBlueprintCardProps> = ({
	thumbnailSrc,
	title,
	byline,
	excerpt,
	previewHref,
	repoHref,
	detailsHref,
}) => (
	<>
		<Card
			className="AtlasBlueprintCard"
			headerIconContainerClassName="AtlasBlueprintCard_HeaderIconContainer"
			headerIconPath={thumbnailSrc}
			contentClassName="AtlasBlueprintCard_Content"
			contentDescriptionClassName="AtlasBlueprintCard_Description"
			contentTitle={title}
			contentSub={byline}
			contentDescription={excerpt}
			content={
				<>
					<Text tag="p">
						<a href={previewHref}>Preview Site</a>
					</Text>
					<Text tag="p">
						<a href={repoHref}>Open the code on Github</a>
					</Text>
					<Divider />
					<Text tag="p">
						<a href={detailsHref}>Show more details</a>
					</Text>
				</>
			}
		/>
	</>
);

export default AtlasBlueprintCard;
