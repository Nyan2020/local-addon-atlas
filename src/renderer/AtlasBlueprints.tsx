import React from 'react';
import {
	Container,
	Title,
	Text,
	TextButton,
} from '@getflywheel/local-components';
import path from 'path';

const atlasBlueprints = [
	{
		thumbnail: path.resolve(
			__dirname,
			'../atlas-blueprints/basic-blueprint.png',
		),
		title: 'Basic Blueprint',
		byline: 'Headless WordPress',
		excerpt:
			'A bare-bones headless site with enough scaffolding to get you started.',
		previewHref: 'http://example.com',
		repoHref: 'http://example.com',
		detailsHref: 'http://example.com',
	},
	{
		thumbnail: path.resolve(
			__dirname,
			'../atlas-blueprints/blog-blueprint.png',
		),

		title: 'Blog Blueprint',
		byline: 'Headless WordPress',
		excerpt: 'A standard blog built to run on headless WordPress.',
		previewHref: 'http://example.com',
		repoHref: 'http://example.com',
		detailsHref: 'http://example.com',
	},
	{
		thumbnail: path.resolve(
			__dirname,
			'../atlas-blueprints/portfolio-blueprint.png',
		),
		title: 'Portfolio Blueprint',
		byline: 'Headless WordPress',
		excerpt:
			'Showcase your skills and blog about them with this Blueprint.',
		previewHref: 'https://example.com',
		repoHref: 'https://example.com',
		detailsHref: 'https://example.com',
	},
];

const AtlasBlueprints: React.FC = () => (
	<Container margin="l" className="AtlasBlueprints">
		<Title size="l" tag="h2">
			Atlas Bluprints
		</Title>
		<Text tag="p">
			Blueprints come with starter code, plugins, content models, and page
			structure to get your Atlas app off the ground faster.{' '}
			<TextButton
				privateOptions={{
					textDecoration: 'underline',
				}}
			>
				Learn more
			</TextButton>
		</Text>
		<Container className="AtlasBlueprintList">
			{atlasBlueprints.map((bp) => (
				<Container className="AtlasBlueprint">
					<img
						src={bp.thumbnail}
						alt={`Preview of the ${bp.title}`}
					/>
					<Container margin="m">
						{' '}
						<Title>{bp.title}</Title>
						<Text>{bp.byline}</Text>
						<Container marginBottom="l">
							<Text tag="p">{bp.excerpt}</Text>
						</Container>
						<Text tag="p">
							<a href={bp.links.preview.href}>
								{bp.links.preview.text}
							</a>
						</Text>
						<Text tag="p">
							<a href={bp.links.repo.href}>
								{bp.links.repo.text}
							</a>
						</Text>
						<Divider />
						<Text tag="p">
							<a href={bp.links.details.href}>
								{bp.links.details.text}
							</a>
						</Text>
					</Container>{' '}
				</Container>
			))}
		</Container>
	</Container>
);

export default AtlasBlueprints;
