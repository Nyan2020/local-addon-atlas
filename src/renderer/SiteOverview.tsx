import React from 'react';
import { TableListRow, TextButton } from '@getflywheel/local-components';
import { SITE_STATUS } from '../constants';

interface Props {
	localUrl: string;
	siteStatus: string;
}

const getSiteStatus = siteStatus => {
	const STATUS = {
		STOPPING: 'Stopping',
		STOPPED: 'Stopped',
		STARTING: 'Starting',
		STARTED: 'Started',
	};

	switch (siteStatus) {
		case SITE_STATUS.STOPPING:
			return STATUS.STOPPING;
		case SITE_STATUS.HALTED:
			return STATUS.STOPPED;
		case SITE_STATUS.RUNNING:
			return STATUS.STARTED;
		case SITE_STATUS.STARTING:
			return STATUS.STARTING;
		default:
			return STATUS.STOPPED;
	}
};

const showTerminalOutput = () => console.log('COMING SOON =)');


const SiteOverview = (props: Props) => {
	const { localUrl, siteStatus } = props;

	return (
		<>
			<TableListRow label="Status" selectable>
				<div style={{ flex: 1, display: 'flex', alignContent: 'center' }}>
					<p>{getSiteStatus(siteStatus)}</p>
					<TextButton size="tiny" onClick={() => showTerminalOutput()}>
						Show Output
					</TextButton>
				</div>
			</TableListRow>
			<TableListRow
				label="Node.js host"
				selectable
			>
				<a href={`http://${localUrl}`}>{localUrl}</a>

			</TableListRow>
		</>
	);
};

export default SiteOverview;
