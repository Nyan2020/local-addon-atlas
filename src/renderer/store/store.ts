import { useSelector, TypedUseSelectorHook } from 'react-redux';
import { createSlice, PayloadAction, current, configureStore } from '@reduxjs/toolkit';
import { SiteEnvironmentData } from '../../types';

export { selectors } from './selectors';

export const addHeadlessEnvironmentSlice = createSlice({
	name: 'addHeadlessEnvironment',
	initialState: {} as SiteEnvironmentData,
	reducers: {
		addHeadlessEnvironment: (state, action: PayloadAction<{siteName: string, isChecked: boolean}>) => {
			state[action.payload.siteName] = action.payload.isChecked;
			return state;
		},
	},
});

export const actions = {
	...addHeadlessEnvironmentSlice.actions,
};

export const store = configureStore({
	reducer: {
		addHeadlessEnvironment: addHeadlessEnvironmentSlice.reducer,
	},
});

export type State = ReturnType<typeof store.getState>;

export const useStoreSelector = useSelector as TypedUseSelectorHook<State>;
