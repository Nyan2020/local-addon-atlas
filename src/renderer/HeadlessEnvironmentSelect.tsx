import React, { useState, useEffect } from 'react';
import { Checkbox, Text } from '@getflywheel/local-components';
import { faustJsDocsUrl } from '../renderer';
import { useObserver } from 'mobx-react';
import { $offline } from '@getflywheel/local/renderer';

export const HeadlessEnvironmentSelect = (props) => useObserver(() => {
	const [checked, setChecked] = useState(false);

	useEffect(() => {
		const disable = $offline.offline && checked;

		props.hooks.addFilter('NewSiteEnvironment_ContinueButton', function () {
			if (disable) {
				return this.renderOfflineButtonWithTooltip();
			}

			return this.renderContinueButton();
		});

		props.rerenderParent();
	}, [checked]);

	return (
		<div>
			<div className="FormRow FormRow__Center AtlasCheckboxSelect" style={{ marginTop: 30 }}>
				<div>
					<Checkbox
						name="useAtlasFramework"
						style={{ marginTop: 10 }}
						checked={checked}
						label="Enable Atlas Add-on on this site."
						onChange={(checked) => setChecked(checked)}
					/>
					<div className="AtlasTextLink">
						<Text>
							(Site will be built with a Node.js front-end powered <br/>
							by <a href={faustJsDocsUrl} style={{ fontWeight: 'bold' }}>WP Engine's Headless WordPress framework, Faust.js</a>)
						</Text>
					</div>
				</div>
			</div>
		</div>
	);
});
