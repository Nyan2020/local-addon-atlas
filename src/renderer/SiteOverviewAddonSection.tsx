import React from 'react';
import { TableList, TableListRow, TextButton } from '@getflywheel/local-components';
import * as LocalRenderer from '@getflywheel/local/renderer';
import { IPC_EVENTS } from './../constants';
import type { Site } from '@getflywheel/local';

interface Props {
	localUrl: string;
	siteStatus: string;
	site: Site;
}

const showTerminalOutput = (site: Site) => LocalRenderer.ipcAsync(
	IPC_EVENTS.OPEN_XTERM,
	site,
);

const renderLocalUrlHyperlink = (isSiteRunning: boolean, localUrl: string) => {
	if (isSiteRunning) {
		return (
			<a href={`http://${localUrl}`}>{localUrl}</a>
		);
	}

	return (<p>{localUrl}</p>);
};

const SiteOverview = (props: Props) => {
	const { localUrl, siteStatus, site } = props;
	const isSiteRunning = siteStatus === 'running';

	return (
		<TableList>
			<TableListRow label="Status" selectable>
				<div style={{ flex: 1, display: 'flex', alignContent: 'center' }}>
					<p style={{ textTransform: 'capitalize' }}>{siteStatus}</p>
					<TextButton
						size="tiny"
						onClick={() => showTerminalOutput(site)}
						disabled={!isSiteRunning}
					>
						Show Output
					</TextButton>
				</div>
			</TableListRow>
			<TableListRow
				label="Node.js host"
				selectable
			>
				{renderLocalUrlHyperlink(isSiteRunning, localUrl)}
			</TableListRow>
		</TableList>
	);
};

export default SiteOverview;
