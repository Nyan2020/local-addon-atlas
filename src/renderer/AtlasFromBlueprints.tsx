import React from 'react';
import {
	Container,
	Title,
	Text,
	TextButton,
	RadioBlock,
} from '@getflywheel/local-components';
import AtlasBlueprintCard from './AtlasBlueprintCard';
import atlasBlueprints from '../../atlas-blueprints/blueprintsContent';

interface IProps {
	setBlueprintOption: (option) => void;
}

const AtlasFromBlueprints: React.FC<IProps> = (props) => {
	const result = atlasBlueprints.reduce((prev, current) => ({
		...prev,
		[current.title]: {
			key: current.repoHref,
			content:
				<AtlasBlueprintCard
					key={current.title}
					thumbnailSrc={current.thumbnail}
					title={current.title}
					byline={current.byline}
					excerpt={current.excerpt}
					previewHref={current.previewHref}
					repoHref={current.repoHref}
					detailsHref={current.detailsHref}
					style={{ height: '100%', margin: '0' }}
				/>,
			style: {
				height: '100%',
				marginRight: '15px',
				textAlign: 'unset',
			},
		},
	}), {});

	const onChange = (option: string) => {
		props.setBlueprintOption({ 'bpName': option, 'atlasUrl': result[option].key });
	};

	return (
		<Container margin="l" className="AtlasBlueprints">
			<Title size="l" tag="h2">
				Atlas Blueprints
			</Title>
			<Container
				marginTop="m"
				marginBottom="l"
				className="AtlasBlueprintsDescription"
			>
				<Text tag="p">
					Blueprints come with starter code, plugins, content models, and
					page structure to get your Atlas app off the ground faster.
					<br />
					<TextButton
						tag="a"
						tagProps={{ href: 'https://localwp.com/help-docs' }}
						privateOptions={{
							textDecoration: 'underline',
						}}
					>
						Learn more about headless WordPress development
					</TextButton>
				</Text>
			</Container>
			<Container className="AtlasBlueprintList">
				<RadioBlock
					onChange={(option) => onChange(option)}
					heightSize='none'
					options={result}
				/>
			</Container>
		</Container>
	);
};

export default AtlasFromBlueprints;
