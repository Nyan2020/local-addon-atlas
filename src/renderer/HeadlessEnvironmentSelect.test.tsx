
/**
 * @jest-environment
 */

import React from 'react';
import Enzyme, { mount } from 'enzyme';
import { Checkbox } from '@getflywheel/local-components';
import { HeadlessEnvironmentSelect } from './HeadlessEnvironmentSelect';
import { Provider } from 'react-redux';
import { store } from './store/store';
import Adapter from 'enzyme-adapter-react-16';
import * as LocalRenderer from '@getflywheel/local/renderer';

Enzyme.configure({ adapter: new Adapter() });

describe('HeadlessEnvironmentSelect', () => {
	const props = {
		siteInfo: { siteName: 'Jacks Flapjacks' },
	};

	it('renders a </Checkbox> with correct label text', () => {
		const componentNode = mount(
			<Provider store={store}>
				<HeadlessEnvironmentSelect
					/* eslint-disable react/jsx-props-no-spreading */
					{...props}
					/* eslint-enable react/jsx-props-no-spreading */
				/>,
			</Provider>,
		);

		const checkboxLabelText = 'Use Atlas framework to build this site headless (beta)';

		const checkbox = componentNode.find(Checkbox);

		expect(checkbox).toBeTruthy();

		expect(checkbox.props().label).toBe(checkboxLabelText);

		expect(componentNode).toMatchSnapshot();
	});

	it('correctly updates state when checked', () => {
		const componentNode = mount(
			<Provider store={store}>
				<HeadlessEnvironmentSelect
					/* eslint-disable react/jsx-props-no-spreading */
					{...props}
					/* eslint-enable react/jsx-props-no-spreading */
				/>,
			</Provider>,
		);

		const checkbox = componentNode.find('input');

		const { siteName } = props.siteInfo;

		checkbox.simulate('change', { checked: true });

		const state = componentNode.props().store.getState();

		expect(state.addHeadlessEnvironment[siteName]).toBeTruthy();

		expect(LocalRenderer.ipcAsync).toHaveBeenCalledTimes(1);
	});
});
