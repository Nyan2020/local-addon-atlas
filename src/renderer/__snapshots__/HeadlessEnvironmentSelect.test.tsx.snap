// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeadlessEnvironmentSelect renders a </Checkbox> with correct label text 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Component
    siteInfo={
      Object {
        "siteName": "Jacks Flapjacks",
      }
    }
  >
    <div
      className="FormRow FormRow__Center AtlasCheckboxSelect"
      style={
        Object {
          "marginTop": 30,
        }
      }
    >
      <div>
        <_
          checked={false}
          disabled={false}
          label="Use Atlas framework to build this site headless (beta)"
          onChange={[Function]}
          style={
            Object {
              "marginTop": 10,
            }
          }
        >
          <div
            className="Checkbox_1GDEs_v15-0-1"
            style={
              Object {
                "marginTop": 10,
              }
            }
          >
            <label
              className="Checkbox_Label_23ova_v15-0-1"
            >
              <input
                checked={false}
                className="Checkbox_InputHidden_2NZeg_v15-0-1"
                disabled={false}
                onChange={[Function]}
                type="checkbox"
              />
              <div
                className="Checkbox_GraphicsCont_1NHXT_v15-0-1"
              />
              <div
                className="Checkbox_LabelContents_UOdqj_v15-0-1"
              >
                Use Atlas framework to build this site headless (beta)
              </div>
            </label>
          </div>
        </_>
      </div>
    </div>
  </Component>
  ,
</Provider>
`;
